### PR TITLE
Fix negative GW storage and flux 

### PR DESCRIPTION
**Pull request by lcunha0118**
This PR fixex the possibility of negative groundwater storage and flow. Included a warning when GW flow is larger than GW storage. 

Other changes include:

- remove soil_params.mult from configuration file and code. This parameter referred to the LKSATFAC in the NWM, but is not used in CFE;
- Removed previously commented Xinanjiang_partitioning_scheme function. Only slightly differences to the function that is currently being used;
- Included Evapotranspiration in the water balance. 

## Testing

1. Testing was performed in the standalone version and on the framework. 

## Notes

- A new PR will be submitted later this week for the GIUH component "/params/". The current function uses files from the old version of the hydrofabric. The function will be updated to use the new files. 

## Todos

-

## Checklist

- [x ] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [ x] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ x] Passes all existing automated tests
- [x ] Any _change_ in functionality is tested
- [ x] New functions are documented (with a description, list of inputs, and expected output)
- [x ] Placeholder code is flagged / future todos are captured in comments
- [x ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

File: src/cfe.c [hunk 1]
```diff
@@ -180,7 +185,15 @@ extern void cfe(
   
   conceptual_reservoir_flux_calc(gw_reservoir_struct,&primary_flux,&secondary_flux);
   
+  
   flux_from_deep_gw_to_chan_m=primary_flux;  // m/h   <<<<<<<<<< BASE FLOW FLUX >>>>>>>>>
+  if(flux_from_deep_gw_to_chan_m >  gw_reservoir_struct->storage_m)  {
+  flux_from_deep_gw_to_chan_m=gw_reservoir_struct->storage_m;
```
Comment (hellkite500) [inline src/cfe.c:191]: What happens to the excess flux (`gw_reservoir_struct->storage_m - flux_from_deep_gw_to_chan_m`)?

File: src/cfe.c [hunk 2]
```diff
@@ -558,111 +577,7 @@ void Xinanjiang_partitioning_scheme(double water_input_depth_m, double field_cap
 
 return;
 }
-//##############################################################
-//########   XINANJIANG RUNOFF PARTITIONING SCHEME   ###########
-/*##############################################################
-
-void Xinanjiang_partitioning_scheme(double water_input_depth_m, double field_capacity_m,
```
Comment (hellkite500) [inline src/cfe.c:565]: Is this function being deprecated?  If so, I would do that in a different PR.

File: src/cfe.c [hunk 3]
```diff
@@ -682,12 +597,13 @@ void et_from_rainfall(double *timestep_rainfall_input_m, struct evapotranspirati
         }
 
         else{
-
-            et_struct->potential_et_m_per_timestep -= *timestep_rainfall_input_m;
+            // LKC: This was incorrectly set to potential instead of actual
+	    et_struct->actual_et_m_per_timestep = *timestep_rainfall_input_m;
```
Comment (hellkite500) [inline src/cfe.c:601]: Just be sure is this intended to be a copy of `timestep_rainfall_input_m`?  The previously incorrect `potential` variable looked like an accumulation using `-=`, where as now there is no cumulative effect, just a straight assignment using `=`.

**Review by hellkite500 — COMMENTED** [review]
Couple of general comments, it looks like this PR contains several changes not all related strictly to `Fix negative GW storage and flux`.  Keeping changes isolated helps in maintaining history as well as expediting reviews.

**Comment by madMatchstick** [comment]
@lcunha0118 - to Nel's point, I believe the only commit you need for this PR is [this one](https://github.com/NOAA-OWP/cfe/pull/43/commits/954d6ffe0487d20f25c5a263e2134e7f587aaa9e). 

Also - If removing `soil_params.mult`, go ahead and remove it's "set check",`is_soil_params__mult_set`, [here](https://github.com/lcunha0118/cfe/blob/0fdd96c8a70fc15169c86e5a6eaa82c8f2ae1b3c/src/bmi_cfe.c#L372).

File: src/cfe.c (hunk 3 above)
Comment (lcunha0118) [reply src/cfe.c:601]: Yes, it was intended to copy timestep_rainfall_input_m. The previous statement with -= was moved outside of the loop to line 606 since it is valid for both conditions.

**Comment by lcunha0118** [comment]
@madMatchstick I removed the is_soil_params__mult_set.
@hellkite500 Sorry for the accumulated commits. I was not able to send just a few of them without creating conflicts. I am new to GitHub so I am not exactly sure why.

File: src/cfe.c (hunk 1 above)
Comment (lcunha0118) [reply src/cfe.c:191]: Nothing has to happen to it. We cannot remove more than it is available in the storage, even if the equation that calculates the flux (which is conceptual) tells us to. So we end up removing ONLY what is available. That should not create any issues with water balance.

Comment (hellkite500) [reply src/cfe.c:191]: My point is _something_ had happened to it. My only recommendation is to add an `error` accumulator to put these types of errors into so you can see them along side other diagnostic info.

**Comment by hellkite500** [comment]
Would be happy to help show some ways to clean up and isolate changes. It is mostly just `git` commands in the right order and some branch management.

File: src/cfe.c (hunk 1 above)
Comment (lcunha0118) [reply src/cfe.c:191]: The flux in line 189 is like a "potential flux" not a real flux. It is calculated based on a conceptual equation (you can even think of it like a "regression"). This flux is the real flux if enough water is available in the reservoir. if not, then the real flux is limited to the amount of water in the storage (line 191). So it is not an error. It does not interfere with the water balance. There is nothing that physically obligates the flux to be equal to what was calculated by the conceptual equation.

Comment (hellkite500) [reply src/cfe.c:191]: Gotcha. It was a little hard to see just based on the diff I was reading. But that is why reviews are useful, just in case and just to learn a little 😁

**Comment by hellkite500** [comment]
Steps to clean up, in a temporary location on your system
```sh
git clone https://github.com/lcunha0118/cfe.git
git remote add upstream https://github.com/noaa-owp/cfe
git checkout -b tmp
git log
```
Find the commit you want use for your PR, and copy the commit hash, e.g.
954d6ffe0487d20f25c5a263e2134e7f587aaa9e
Now get the latest upstream master branch and update your master so they are at the same commit
```sh
git fetch upstream
git reset --hard upstream/master
```
Now your fork/branch is even with master.
Create a branch from this point to contain the updates you want to PR.  Form this branch, we will steal the commit we want to use using cherry pick
```sh
git checkout -b fix-neg-gw
git cherry-pick 954d6ffe0487d20f25c5a263e2134e7f587aaa9e
```
If you need to adjust the changes in this commit, you can reset and re-commit:
```sh
git reset --soft HEAD~2
git status
```
And you will see the various files in that commit.  You can then use
```sh
git add include/cfe.h src/cfe.c
git commit -m "fix negative GW storage and flux"
```
To stage just the required files for the fix.  Finally, we can either update the PR branch (in this case master on a fork), or push our new branch and pr that one.
To update the current PR
```sh
git push origin fix-neg-gw:master
```
You may need to force this change using `-f`
```sh
git push -f origin fix-neg-gw:master
```
Or to push the branch and open a new PR
```sh
git push origin fix-neg-gw
```
Then when you go to the repository on GitHub you should see a prompt to open a PR with this newly pushed branch.

**Comment by madMatchstick** [comment]
Okay to merge.

**Review by madMatchstick — APPROVED** [review]
good to merge
